docs/_component/editor.client.js

```diff
@@ -0,0 +1,215 @@
+import {VFileMessage} from 'vfile-message'
+import {statistics} from 'vfile-statistics'
+import {reporter} from 'vfile-reporter'
+import {evaluate} from 'xdm'
+import CodeMirror from 'rodemirror'
+import {basicSetup} from '@codemirror/basic-setup'
+import {markdown as langMarkdown} from '@codemirror/lang-markdown'
+import {oneDark} from '@codemirror/theme-one-dark'
+import {Tab, Tabs, TabList, TabPanel} from 'react-tabs'
+import {ErrorBoundary} from 'react-error-boundary'
+import {toH} from 'hast-to-hyperscript'
+import {lowlight} from 'lowlight/lib/core.js'
+import javascript from 'highlight.js/lib/languages/javascript.js'
+import json from 'highlight.js/lib/languages/json.js'
+import markdown from 'highlight.js/lib/languages/markdown.js'
+
+lowlight.registerLanguage('js', javascript)
+lowlight.registerLanguage('json', json)
+lowlight.registerLanguage('md', markdown)
+
+function useXdm() {
+  const [file, setFile] = useState(null)
+
+  const setValue = useCallback(async value => {
+    const file = new VFile({basename: 'example.mdx', value})
+
+    const capture = name => () => tree => {
+      file.data[name] = tree
+    }
+
+    try {
+      file.result = (
+        await evaluate(file, {
+          ...runtime,
+          remarkPlugins: [capture('mdast')],
+          rehypePlugins: [capture('hast')],
+          recmaPlugins: [capture('esast')]
+        })
+      ).default
+    } catch (error) {
+      const message =
+        error instanceof VFileMessage ? error : new VFileMessage(error)
+
+      if (!file.messages.includes(message)) {
+        file.messages.push(message)
+      }
+
+      message.fatal = true
+    }
+
+    setFile(file)
+  }, [])
+
+  return [file, setValue]
+}
+
+export function Editor({children}) {
+  const defaultValue = children
+  const extensions = useMemo(() => [basicSetup, oneDark, langMarkdown()], [])
+  const [file, setValue] = useXdm()
+  const stats = file ? statistics(file) : {}
+
+  return (
+    <div className="editor">
+      <div className="editor-input">
+        <noscript>
+          <pre>
+            <code className="hljs language-md">
+              {toH(createElement, lowlight.highlight('md', defaultValue))}
+            </code>
+          </pre>
+        </noscript>
+        <CodeMirror
+          value={defaultValue}
+          extensions={extensions}
+          onUpdate={v => {
+            if (v.docChanged) {
+              setValue(v.state.doc.toString())
+            }
+          }}
+        />
+      </div>
+      <Tabs className="editor-result">
+        <div className="editor-result-tabs">
+          <TabList className="editor-result-tabs-scroll">
+            {[
+              'Run',
+              'Compile',
+              'mdast (markdown)',
+              'hast (HTML)',
+              'esast (JS)'
+            ].map(label => {
+              const compile = label === 'Compile'
+              const className =
+                'editor-result-tab' + (compile ? ' editor-result-badge' : '')
```

Comment:
it is optional, but often nice to have a utility like https://www.npmjs.com/package/clsx to help manage dynamic classname/string-property construction.